### ISSUE TITLE
Support for different runtimes

### ISSUE THREAD

**Issue by kvz** [transloadit/node-sdk#199]
For example:

- [Cloudflare Workers](https://developers.cloudflare.com/workers/)
- [Vercel functions](https://vercel.com/docs/functions)
- [AWS Lambda](https://aws.amazon.com/lambda/)
- Deno
- Bun

And maybe even

- Browsers
- React Native

We should then rename the node-sdk to `@transloadit/client` perhaps. Optionally we can support different runtimes with named exports like `import { BunClient } from '@transloadit/client'` or similar.

**Comment by mifi** [comment]
In order to support all these platforms with upload progress we could either:

Make our own abstraction above fetch and XMLHttpRequest (fetch is needed for Node.js and React Native, XMLHttpRequest is needed for upload progress).

Use a library: The only popular library that I can find that satisfies all of these is the good ol' `axios`, but it has a relatively large bundle size of 13.7kB gzipped. I tried asking AI but they just hallucinate and also cannot find any libraries other than axios that satisfy all the requirements:

- upload progress support
- runs in browsers (xmlhttprequest for upload progress)
- runs in react native (I believe fetch API)
- runs in node.js

Alternatively just use `fetch`, but then the only browser that we would be able to support [upload progress in is chrome](https://stackoverflow.com/questions/35711724/upload-progress-indicators-for-fetch/69400632#69400632).

**Comment by Acconut** [comment]
> * upload progress support

It's possible to get upload progress using the Fetch API by jumping through a few hoops: https://x.com/vvoyer/status/1856287377188368449

At some point Fetch will get native support for upload progress, but not soon: https://github.com/whatwg/fetch/issues/607

**Comment by mifi** [comment]
> It's possible to get upload progress using the Fetch API by jumping through a few hoops: https://x.com/vvoyer/status/1856287377188368449

afaik this is what i mentioned above, that this workaround only works for chrome, and the only alternative that works cross-browser is XMLHttprequest

**Comment by Acconut** [comment]
> that this workaround only works for chrome

You're right, thanks! I didn't know this but the tweet also means the lacking support in Firefox and Safari. That's unfortunate. Upload progress is an important components for good user experience and ensuring its always available is also a hassle in tus-js-client.

**Comment by mifi** [comment]
Yup. If we want to support all these platforms (or specifically node.js+browser) **and** support upload progress, I think it boils down to that we have to either:
1. just use axios, or
2. make a wrapper around fetch and XMLHttpRequest which supports upload progress, or maybe we can even make a new open source http request library a.la `ky` that does this, because evidently there's lack of such a library in the npm ecosystem

**Comment by mifi** [comment]
> That being said, I am wondering if upload progress is a concern in the Node SDK itself. Isn't the uploading of files deferred to tus-js-client nowadays? If so, we have to think about upload progress in tus-js-client. That would likely be a similar discussion, but Node SDK could already move to the Fetch API and we just have to figure out how to make tus-js-client compatible with these platforms while still offering upload progress. Does that make sense to you?

That's a good point. tus-js-client is being used by node-sdk iff `isResumable: true` and the value false is deprecated so maybe we should remove it in the next major? https://github.com/transloadit/node-sdk/blob/3c0ef2ea8ed23dbc79ec5540a3349aa3c6df4856/src/Transloadit.ts#L175

like you said, if we remove the option it will allow us to only use tus-js-client for uploads, and thus we can use **only** fetch because we don't need any upload progress in node-sdk. wdyt @kvz can we remove the deprecated `isResumable` option in the next major?

**Comment by kvz** [comment]
Ah great idea to bundle that removal in the next major, just in time!

I think we should also bundle the robot types in the major @remcohaszing, we’re so close 🤞

**Comment by kvz** [comment]
We're now a multi module monorepo here, with `@transloadit/node` (and `transloadit` as an alias) `@transloadit/types` `@transloadit/utils` `@transloadit/zod/v3` `@transloadit/zod/v4`. This opens the door to a  `@transloadit/browser` (that Uppy can then use), and all the things discussed in this thread.
